src/NServiceBus.Gateway/Gateway.cs

```diff
@@ -54,29 +54,30 @@ protected override void Setup(FeatureConfigurationContext context)
 
             RegisterChannels(context, channelManager, out var channelSenderFactory, out var channelReceiverFactory);
 
-            var logicalAddress = new QueueAddress(context.Settings.EndpointQueueName(), null, null, "gateway");
-            var gatewayInputAddress = transportDefinition.ToTransportAddress(logicalAddress);
-
             var retryPolicy = context.Settings.Get<Func<IncomingMessage, Exception, int, TimeSpan>>("Gateway.Retries.RetryPolicy");
 
+            var logicalGatewayAddress = new QueueAddress(context.Settings.EndpointQueueName(), null, null, "gateway");
             var replyToAddress = GetReplyToAddress(context.Settings, channelManager);
 
-            var sender = new GatewayMessageSender(
-                gatewayInputAddress,
+            context.Services.AddSingleton(b => new GatewayMessageSender(
+                b.GetService<ITransportAddressResolver>().ToTransportAddress(logicalGatewayAddress),
                 new MessageNotifier(),
-                context.Settings.LocalAddress(),
+                b.GetService<ReceiveAddresses>().MainReceiveAddress,
```

Comment:
can the gateway somehow be run in send-only mode that would fail to resolve the service?

Reply:
Not sure but since the gateway is using satellites to do the forwarding you can't use it if the endpoint is sendonly?

Unless there is an option to run satellites but still be send only?

Reply:
And the previous call to context.Settings.LocalAddress() would have thrown on a sendonly endpoint

Reply:
> Unless there is an option to run satellites but still be send only?

I don't think there is an option but I think this fails silently

> And the previous call to context.Settings.LocalAddress() would have thrown on a sendonly endpoint

I don't think that's true? https://github.com/Particular/NServiceBus/blob/7.5.0/src/NServiceBus.Core/SettingsExtensions.cs#L57 does not check for send-only?

Reply:
So the exception isn't correct then? https://github.com/Particular/NServiceBus/blob/7.5.0/src/NServiceBus.Core/SettingsExtensions.cs#L54